File: matching-engine/src/main/java/com/titanx/persistence/RedisPositionCache.java
package com.titanx.persistence;

import io.lettuce.core.RedisClient;
import io.lettuce.core.api.StatefulRedisConnection;
import io.lettuce.core.api.sync.RedisCommands;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

/**
 * ============================================================
 * TitanX RedisPositionCache — Phase 3: Hot-Path State
 * ============================================================
 *
 * Redis serves as the hot-path store for two things:
 *
 * 1. Open order state:   "is orderId X still active?" → O(1) lookup
 * 2. Account positions:  real-time net position per account per symbol
 *
 * Why Redis and not the database?
 *  "PostgreSQL P99 is ~5ms for a write. Redis P99 is ~0.1ms.
 *   On the matching hot path (100K orders/sec), we cannot afford 5ms
 *   per position check. Redis gives us the speed with acceptable
 *   durability (AOF persistence, replica failover)."
 *
 * Redis key schema:
 *   pos:{accountId}:{symbol}   → net position (integer, signed)
 *   order:{orderId}:status     → ORDER_STATUS enum string
 *   pubsub channel: fills:{symbol}
 */
public class RedisPositionCache {

    private static final Logger log = LoggerFactory.getLogger(RedisPositionCache.class);

    private final RedisClient client;
    private final StatefulRedisConnection<String, String> connection;
    private final RedisCommands<String, String> commands;

    public RedisPositionCache(String redisUri) {
        this.client = RedisClient.create(redisUri);
        this.connection = client.connect();
        this.commands = connection.sync();
        log.info("RedisPositionCache connected to {}", redisUri);
    }

    // ---- Position tracking ----

    public void updatePosition(String accountId, String symbol, long delta) {
        String key = positionKey(accountId, symbol);
        commands.incrby(key, delta);
    }

    public long getPosition(String accountId, String symbol) {
        String val = commands.get(positionKey(accountId, symbol));
        return val == null ? 0L : Long.parseLong(val);
    }

    // ---- Order status cache ----

    public void setOrderStatus(String orderId, String status) {
        commands.setex(orderKey(orderId), 3600, status); // 1hr TTL
    }

    public String getOrderStatus(String orderId) {
        return commands.get(orderKey(orderId));
    }

    // ---- Pub/Sub fill broadcast ----

    /**
     * Publish a fill notification to all WebSocket subscribers.
     * Channel: fills:{symbol}
     * Message: JSON payload of the trade
     */
    public void publishFill(String symbol, String tradeJson) {
        long subscribers = commands.publish("fills:" + symbol, tradeJson);
        log.debug("Published fill to fills:{}, {} subscribers", symbol, subscribers);
    }

    // ---- Flush / Admin ----

    public Map<String, String> getPositionSummary(String accountId) {
        // In production: use HSCAN over a hash key per account
        // Simplified: assume symbols are known
        return Map.of();
    }

    public void close() {
        connection.close();
        client.shutdown();
    }

    private String positionKey(String accountId, String symbol) {
        return "pos:" + accountId + ":" + symbol;
    }

    private String orderKey(String orderId) {
        return "order:" + orderId + ":status";
    }
}

File: backend/src/persistence/RedisPositionCache.cpp
#include "titanx/persistence/RedisPositionCache.hpp"
#include <hiredis/hiredis.h>
#include <spdlog/spdlog.h>
#include <stdexcept>

namespace titanx {

RedisPositionCache::RedisPositionCache(const std::string& host, int port)
    : ctx_(nullptr) {

    struct timeval timeout = {1, 500000}; // 1.5 seconds
    ctx_ = redisConnectWithTimeout(host.c_str(), port, timeout);

    if (ctx_ == nullptr || ctx_->err) {
        std::string err = ctx_ ? ctx_->errstr : "allocation failed";
        spdlog::error("RedisPositionCache: connection failed: {}", err);
        if (ctx_) { redisFree(ctx_); ctx_ = nullptr; }
        // Don't throw — allow graceful degradation
        return;
    }
    spdlog::info("RedisPositionCache connected to {}:{}", host, port);
}

RedisPositionCache::~RedisPositionCache() {
    close();
}

// ---- Position tracking ----

void RedisPositionCache::updatePosition(const std::string& accountId,
                                        const std::string& symbol,
                                        int64_t delta) {
    if (!ctx_) return;
    std::string key = positionKey(accountId, symbol);
    redisReply* reply = static_cast<redisReply*>(
        redisCommand(ctx_, "INCRBY %s %lld", key.c_str(), (long long)delta));
    if (reply) freeReplyObject(reply);
}

int64_t RedisPositionCache::getPosition(const std::string& accountId,
                                        const std::string& symbol) {
    if (!ctx_) return 0;
    std::string key = positionKey(accountId, symbol);
    redisReply* reply = static_cast<redisReply*>(
        redisCommand(ctx_, "GET %s", key.c_str()));

    int64_t result = 0;
    if (reply) {
        if (reply->type == REDIS_REPLY_STRING && reply->str) {
            result = std::stoll(reply->str);
        }
        freeReplyObject(reply);
    }
    return result;
}

// ---- Order status cache ----

void RedisPositionCache::setOrderStatus(const std::string& orderId,
                                        const std::string& status) {
    if (!ctx_) return;
    std::string key = orderKey(orderId);
    redisReply* reply = static_cast<redisReply*>(
        redisCommand(ctx_, "SETEX %s 3600 %s", key.c_str(), status.c_str()));
    if (reply) freeReplyObject(reply);
}

std::string RedisPositionCache::getOrderStatus(const std::string& orderId) {
    if (!ctx_) return "";
    std::string key = orderKey(orderId);
    redisReply* reply = static_cast<redisReply*>(
        redisCommand(ctx_, "GET %s", key.c_str()));

    std::string result;
    if (reply) {
        if (reply->type == REDIS_REPLY_STRING && reply->str) {
            result = reply->str;
        }
        freeReplyObject(reply);
    }
    return result;
}

// ---- Pub/sub fill broadcast ----

void RedisPositionCache::publishFill(const std::string& symbol,
                                     const std::string& tradeJson) {
    if (!ctx_) return;
    std::string channel = "fills:" + symbol;
    redisReply* reply = static_cast<redisReply*>(
        redisCommand(ctx_, "PUBLISH %s %s", channel.c_str(), tradeJson.c_str()));
    if (reply) {
        spdlog::debug("Published fill to fills:{}, {} subscribers",
                      symbol, reply->integer);
        freeReplyObject(reply);
    }
}

void RedisPositionCache::close() {
    if (ctx_) {
        redisFree(ctx_);
        ctx_ = nullptr;
    }
}

// ---- Key builders ----

std::string RedisPositionCache::positionKey(const std::string& accountId,
                                            const std::string& symbol) const {
    return "pos:" + accountId + ":" + symbol;
}

std::string RedisPositionCache::orderKey(const std::string& orderId) const {
    return "order:" + orderId + ":status";
}

} // namespace titanx
